packages/kbn-test/src/page_load_metrics/navigation.ts

```diff
@@ -0,0 +1,130 @@
+  }`;
+}
+
+export function createUrl(path: string, options: NavigationOptions) {
+  const baseUrl = Url.parse(options.appConfig.url);
+  return Url.format({
+    protocol: baseUrl.protocol,
+    hostname: baseUrl.hostname,
+    port: baseUrl.port,
+    pathname: joinPath(baseUrl.pathname || '', path),
+  });
+}
+
+async function loginToKibana(
+  log: ToolingLog,
+  browser: puppeteer.Browser,
+  options: NavigationOptions
+) {
+  log.debug(`log in to the app..`);
+  const page = await browser.newPage();
+  const loginUrl = createUrl('/login', options);
+  await page.goto(loginUrl, {
+    waitUntil: 'networkidle0',
+  });
+  await page.type('[data-test-subj="loginUsername"]', options.appConfig.username);
+  await page.type('[data-test-subj="loginPassword"]', options.appConfig.password);
+  await page.click('[data-test-subj="loginSubmit"]');
+  await page.waitForNavigation({ waitUntil: 'networkidle0' });
+  await page.close();
+}
+
+export async function navigateToApps(log: ToolingLog, options: NavigationOptions) {
+  const browser = await puppeteer.launch({ headless: options.headless, args: ['--no-sandbox'] });
+  const devToolsResponses: NavigationResults = new Map();
+  const apps = [
+    { path: '/app/discover', locator: '[data-test-subj="discover-sidebar"]' },
+    { path: '/app/home', locator: '[data-test-subj="homeApp"]' },
+    { path: '/app/canvas', locator: '[data-test-subj="create-workpad-button"]' },
+    { path: '/app/maps', locator: '[data-test-subj="newMapLink"]' },
+    { path: '/app/apm', locator: '[data-test-subj="apmMainContainer"]' },
+  ];
+
+  await loginToKibana(log, browser, options);
+
+  await Promise.all(
+    apps.map(async (app) => {
+      const page = await browser.newPage();
+      page.setCacheEnabled(false);
+      page.setDefaultNavigationTimeout(0);
+      const frameResponses = new Map<string, FrameResponse>();
+      devToolsResponses.set(app.path, frameResponses);
+
+      const client = await page.target().createCDPSession();
+      await client.send('Network.enable');
+
+      function getRequestData(requestId: string) {
+        if (!frameResponses.has(requestId)) {
+          frameResponses.set(requestId, { url: '', dataLength: 0 });
+        }
+
+        return frameResponses.get(requestId)!;
+      }
+
+      client.on('Network.responseReceived', (event: ResponseReceivedEvent) => {
+        getRequestData(event.requestId).url = event.response.url;
+      });
+
+      client.on('Network.dataReceived', (event: DataReceivedEvent) => {
+        getRequestData(event.requestId).dataLength += event.dataLength;
+      });
+
+      const url = createUrl(app.path, options);
+      log.debug(`goto ${url}`);
+      await page.goto(url, {
+        waitUntil: 'networkidle0',
+      });
+
+      await page.waitForSelector(app.locator, { timeout: 30000 }).catch((err) => {
+        log.error(`Page '${app.path}' was not loaded properly, current url: ${page.url()}`);
+        throw err;
+      });
```

Comment:
I think it would be nice to log while this is happening, to include the selector, and avoid mixing async/await + `.catch()` calls. What about something like:

```suggestion
      let readyAttempt = 0
      let selectorFound = false
      while (!selectorFound) {
        readyAttempt += 1
        try {
          await page.waitForSelector(app.locator, { timeout: 5000 });
          selectorFound = true
        } catch (error) {
          log.error(`Page '${app.path}' was not loaded properly, unable to find '${app.locator}', url: ${page.url()}`);
          
          if (readyAttempt < 6) {
            continue;
          }
          
          throw err;
        }
      });
```

Reply:
Thank you, Spencer!